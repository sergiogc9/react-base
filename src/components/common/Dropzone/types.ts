import { TProps } from '@src/lib/i18n';

export type AddFileError = {
	error: 'file_too_large' | 'incorrect_file_type'
	file: File
};

export type StateProps = {};

export type DispatchProps = {};

export type OwnProps = {
	files: File[]
	multiple: boolean
	accept?: string
	maxSize?: number
	showError?: boolean
	onFilesAdded(files: File[]): void
	onAddFilesError(errors: AddFileError[]): void
	onRemoveFile(file: File): void
};

export type ComponentProps = StateProps & DispatchProps & OwnProps & TProps;
