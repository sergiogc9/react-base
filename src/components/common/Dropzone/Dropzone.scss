@import "@src/scss/_globals.scss";
@import "@src/scss/_mixins.scss";
@import "@src/scss/_variables.scss";

#dropzoneContainer {
	display: flex;
	align-items: center;
	position: relative;
	width: 30rem;
	max-width: 100%;
	border: 1px dashed;
	border-radius: 3px;
	border-color: $md-grey-400;
	padding: 1.6rem;
	overflow: hidden;
	&::before {
		content: "";
		position: absolute;
		left: 0;
		top: 0;
		width: 0;
		opacity: 0.2;
		transition: width 0.5s ease;
	}
	&.dropzone-error{
		border-color: $red-color;
	}
	.dropzone-inner-container {
		width: 100%;

		.dropzone-buttons {
			.md-file-input-container {
				margin-right: 1.5rem;

				label {
					background: $launchmetrics-green;
					white-space: nowrap;
					overflow: hidden;
					text-overflow: ellipsis;
					color: #fff;
					font-size: 1.4rem;
					font-weight: 500;
					line-height: 2.5rem;
					text-transform: uppercase;
					padding: 0.5rem 0.6rem;
					box-shadow: none;
					.md-icon-text {
						padding: 0;
					}
				}
			}
			.dropzone-help {
				display: inline-flex;
				align-content: center;
				font-size: 1.4rem;
				font-weight: 500;
				line-height: 2.5rem;
				padding: 0.5rem 0.6rem;
				margin: 0 0.3rem;
				color: $md-grey-500;

				.icon-upload {
					display: inline-flex;
					align-items: center;
					color: $md-grey-500;
					font-size: 1.8rem;
					margin-right: 0.8rem;
				}
			}
		}
		#dropzoneUploadedFiles {
			margin-top: 1.5rem;

			.dropzone-upload-files-inner {
				display: flex;
				flex-direction: column;

				.dropzone-uploaded-file {
					display: flex;
					align-items: center;
					margin-bottom: 0.5rem;

					.dropzone-uploaded-file-info {
						display: flex;
						align-items: center;
						flex: 1;
						height: 3.2rem;
						transition: height 0.1s;
						overflow: hidden;
						font-size: 1.3rem;

						.dropzone-uploaded-file-name {
							font-weight: 500;
							white-space: nowrap;
							overflow: hidden;
							text-overflow: ellipsis;
							line-height: 1.6rem;
							color: $md-grey-500;
						}
					}

					.dropzone-uploaded-file-commands {
						display: flex;
						margin: 0 0.5rem;

						.dropzone-uploaded-file-commands-remove {
							width: auto;
							height: auto;
							min-height: 0;
							min-width: 0;
							padding: 0;

							.md-ink-container {
								border-radius: 2rem;
							}

							i {
								font-size: 2rem;
							}

							.md-icon-text {
								padding-left: 0;
							}
						}
					}
				}
			}
		}
	}
	&.highlighted-container {
		border-color: $launchmetrics-green;

		.dropzone-inner-container {
			.dropzone-buttons {
				.dropzone-help {
					color: $launchmetrics-green;

					.icon-upload {
						color: $launchmetrics-green;
					}
				}
			}
		}
		&::before {
			width: 100%;
			height: 100%;
			background: rgb(197, 202, 213);
		}
	}
	&.filled-container {
		.dropzone-inner-container {
			.dropzone-buttons {
				.md-file-input-container {
					label {
						background: $md-grey-300;
						color: $md-grey-500;
					}
				}
			}
		}
	}
}
