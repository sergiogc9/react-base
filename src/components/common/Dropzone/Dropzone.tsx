import React from 'react';
import { withT } from '@src/lib/i18n';
import FileInput from 'react-md/lib/FileInputs/FileInput';
import FontIcon from 'react-md/lib/FontIcons/FontIcon';
import Button from 'react-md/lib/Buttons/Button';
import isEmpty from 'lodash/isEmpty';
import { ComponentProps, AddFileError } from './types';

import './Dropzone.scss';

type ComponentState = {
	highlight: boolean,
	disabled: boolean
};

class Dropzone extends React.Component<ComponentProps, ComponentState> {

	constructor(props: ComponentProps) {
		super(props);
		this.state = {
			highlight: false,
			disabled: false
		};
	}

	private _fileListToArray(list: FileList) {
		const array: File[] = [];
		for (let i = 0; i < list.length; i++) {
			array.push(list.item(i)!);
		}
		return array;
	}

	private _getValidationExpression() {
		const { accept } = this.props;
		if (!accept) return null;

		return new RegExp('^(' + accept.replace(/[, ]+/g, '|').replace(/\/\*/g, '/.*') + ')$', 'i');
	}

	private _onDragOver(event: React.DragEvent) {
		const { disabled } = this.state;

		event.preventDefault();

		if (disabled) return;

		this.setState({ highlight: true });
	}

	private _onDragLeave() {
		this.setState({ highlight: false });
	}

	private _onDrop(event: React.DragEvent) {
		const { disabled } = this.state;
		const { multiple } = this.props;

		event.preventDefault();

		if (disabled) return;

		const droppedFiles = event.dataTransfer.files;
		let filesArray: File[] = [];

		if (multiple) filesArray = this._fileListToArray(droppedFiles);
		else if (droppedFiles.length > 0) filesArray = [droppedFiles.item(0)!];

		this._handleFilesChange(filesArray);
		this.setState({ highlight: false });
	}

	private _onFileInputChange(files: File[] | File) {
		if (!Array.isArray(files)) files = [files];
		this._handleFilesChange(files);
	}

	private _handleFilesChange(files: File[]) {
		const { maxSize } = this.props;
		const { onFilesAdded, onAddFilesError } = this.props;
		const regex = this._getValidationExpression();
		const errors: AddFileError[] = [];
		files.forEach(file => {
			const fileExtension = file.name.match(/\.[^\.]*$|$/)![0];
			if (regex && !(regex.test(file.type) || regex.test(fileExtension))) {
				errors.push({ error: 'incorrect_file_type', file });
				return;
			}

			if (maxSize && file.size > maxSize) {
				errors.push({ error: 'file_too_large', file });
				return;
			}
		});
		if (errors.length > 0) onAddFilesError(errors);
		else onFilesAdded(files);
	}

	private _renderUploadedFilesZone() {
		const { files } = this.props;
		return (
			<div id="dropzoneUploadedFiles">
				<div className="dropzone-upload-files-inner">
					{files.map((file, index) => (
						<div key={`dropzoneUploadedFile_${file.name}_${index}`} className="dropzone-uploaded-file">
							<div className="dropzone-uploaded-file-info">
								<div className="dropzone-uploaded-file-name">
									{file.name}
								</div>
							</div>
							<div className="dropzone-uploaded-file-commands">
								<Button
									id={`dropzoneUploadedFile${index}RemoveButton`}
									flat
									iconChildren="close"
									className="dropzone-uploaded-file-commands-remove"
									children=""
									onClick={() => this.props.onRemoveFile(file)}
								/>
							</div>
						</div>
					))}
				</div>
			</div>
		);
	}

	public componentDidUpdate(prevProps: ComponentProps, prevState: ComponentState, snapshot: any) {
		const { disabled } = this.state;
		const { multiple, files } = this.props;
		if (!disabled && !multiple && !isEmpty(files)) this.setState({ disabled: true });
		else if (disabled && (isEmpty(files) || multiple)) this.setState({ disabled: false });
	}

	public render() {
		const { multiple, files, showError, t } = this.props;
		const { highlight, disabled } = this.state;
		let containerClass = 'dropzone-outer-container';
		if (highlight) containerClass += ' highlighted-container';
		else if (disabled) containerClass += ' filled-container';
		if (showError) containerClass += ' dropzone-error';

		return (
			<div id="dropzoneContainer"
				className={containerClass}
				onDragOver={event => { event.preventDefault(); this._onDragOver(event); }}
				onDragLeave={event => this._onDragLeave()}
				onDrop={event => this._onDrop(event)}
			>
				<div className="dropzone-inner-container">
					<div className="dropzone-buttons">
						<FileInput
							id="dropzoneFileInput"
							label={t('newsletter.digest.edit.custom_image.upload.addFiles.one')}
							onChange={uploadedFiles => this._onFileInputChange(uploadedFiles!)}
							allowDuplicates={true}
							multiple={multiple}
							icon={<></>}
							disabled={disabled}
						></FileInput>
						<div className="dropzone-help">
							<FontIcon className="icon-upload"></FontIcon>
							<span>{t(`newsletter.digest.edit.custom_image.upload.dropFiles.${multiple ? 'many' : 'one'}`)}</span>
						</div>
					</div>
					{!isEmpty(files) ? this._renderUploadedFilesZone() : null}
				</div>
			</div>
		);
	}
}

export default withT(Dropzone);
