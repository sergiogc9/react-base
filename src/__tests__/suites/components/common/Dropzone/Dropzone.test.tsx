import React from "react";
import { shallow } from "enzyme";
import cloneDeep from 'lodash/cloneDeep';

import Dropzone from '@src/components/common/Dropzone/Dropzone';

const date = new Date();

const file = {
	lastModified: date.getTime(),
	name: 'test_file.png',
	size: 30000,
	type: 'image/png',
};

const extraFile = {
	lastModified: date.getTime(),
	name: 'test_extra_file.mp4',
	size: 30000,
	type: 'video/mp4',
};

const mockFileList = {
	files: [file, extraFile],
	length: 2,
	item(index: number) {
		return this.files[index];
	}
};

const event = {
	preventDefault: () => false,
	dataTransfer: { files: mockFileList }
};

describe("Test dropzone component", () => {

	let wrapper: any;
	let dropzone: any;
	let fileInput: any;
	let onFilesAdded: any;
	let onAddFilesError: any;
	let onRemoveFile: any;

	beforeEach(() => {
		onFilesAdded = jest.fn();
		onAddFilesError = jest.fn();
		onRemoveFile = jest.fn();

		wrapper = shallow(
			<Dropzone
				files={[]}
				multiple={false}
				onFilesAdded={onFilesAdded}
				onAddFilesError={onAddFilesError}
				onRemoveFile={onRemoveFile}
			/>
		);
		dropzone = wrapper.find("#dropzoneContainer").at(0);
		fileInput = wrapper.find("#dropzoneFileInput").at(0);
	});

	it("Test on drag over / leave", () => {
		dropzone.props().onDragOver(event);
		expect(wrapper.state('highlight')).toEqual(true);
		dropzone.props().onDragLeave(event);
		expect(wrapper.state('highlight')).toEqual(false);
	});

	it("Test drop one file", () => {
		expect(onFilesAdded).toHaveBeenCalledTimes(0);
		dropzone.props().onDrop(event);
		expect(onFilesAdded).toHaveBeenCalledTimes(1);
		expect(onFilesAdded).toHaveBeenCalledWith([file]);
	});

	it("Test drop multiple files", () => {
		expect(onFilesAdded).toHaveBeenCalledTimes(0);
		wrapper.setProps({ multiple: true });
		dropzone.props().onDrop(event);
		expect(onFilesAdded).toHaveBeenCalledTimes(1);
		expect(onFilesAdded).toHaveBeenCalledWith([file, extraFile]);
	});

	it("Test drop no files", () => {
		const modifiedEvent = cloneDeep(event);
		modifiedEvent.dataTransfer.files = {
			...mockFileList,
			files: [],
			length: 0
		};
		expect(onFilesAdded).toHaveBeenCalledTimes(0);
		dropzone.props().onDrop(modifiedEvent);
		expect(onFilesAdded).toHaveBeenCalledTimes(1);
		expect(onFilesAdded).toHaveBeenCalledWith([]);
	});

	it("Test file input", () => {
		expect(onFilesAdded).toHaveBeenCalledTimes(0);
		fileInput.props().onChange(file);
		expect(onFilesAdded).toHaveBeenCalledTimes(1);
		expect(onFilesAdded).toHaveBeenCalledWith([file]);
	});

	it("Test file input multiple files", () => {
		const fileList = [file, extraFile];
		wrapper.setProps({ multiple: true });

		expect(onFilesAdded).toHaveBeenCalledTimes(0);
		fileInput.at(0).props().onChange(fileList);

		expect(onFilesAdded).toHaveBeenCalledTimes(1);
		expect(onFilesAdded).toHaveBeenCalledWith(fileList);
	});

	it("Test file type validation", () => {
		wrapper.setProps({ accept: 'image/*' });
		dropzone.props().onDrop(event);
		expect(onFilesAdded).toHaveBeenCalledTimes(1);
		expect(onFilesAdded).toHaveBeenCalledWith([file]);

		wrapper.setProps({ accept: 'image/png' });
		dropzone.props().onDrop(event);
		expect(onFilesAdded).toHaveBeenCalledTimes(2);
		expect(onFilesAdded).toHaveBeenCalledWith([file]);

		wrapper.setProps({ accept: '.png' });
		dropzone.props().onDrop(event);
		expect(onFilesAdded).toHaveBeenCalledTimes(3);
		expect(onFilesAdded).toHaveBeenCalledWith([file]);

		wrapper.setProps({ accept: 'video/mp4' });
		dropzone.props().onDrop(event);

		expect(onFilesAdded).toHaveBeenCalledTimes(3);
		expect(onAddFilesError).toHaveBeenCalledTimes(1);
		expect(onAddFilesError).toHaveBeenCalledWith([{ error: 'incorrect_file_type', file }]);
	});

	it("Test file size validation", () => {
		wrapper.setProps({ maxSize: 1000 });
		dropzone.props().onDrop(event);
		expect(onFilesAdded).toHaveBeenCalledTimes(0);
		expect(onAddFilesError).toHaveBeenCalledTimes(1);
		expect(onAddFilesError).toHaveBeenCalledWith([{ error: 'file_too_large', file }]);
	});

	it("Test files render", () => {
		wrapper.setProps({ files: [file, extraFile] });
		expect(wrapper.find('#dropzoneUploadedFiles').html()).toMatchSnapshot();
	});

	it("Test file delete button", () => {
		wrapper.setProps({ files: [file, extraFile] });
		wrapper.find('.dropzone-uploaded-file-commands-remove').at(0).simulate('click');
		expect(onRemoveFile).toBeCalledTimes(1);
		expect(onRemoveFile).toHaveBeenCalledWith(file);
	});

	it("Test disabled behaviour", () => {
		wrapper.setProps({ files: [file] });
		expect(wrapper.state('disabled')).toEqual(true);

		dropzone.props().onDragOver(event);
		expect(wrapper.state('highlight')).toEqual(false);

		dropzone.props().onDrop(event);
		expect(onFilesAdded).toHaveBeenCalledTimes(0);

		wrapper.setProps({ files: [file, extraFile], multiple: true });
		expect(wrapper.state('disabled')).toEqual(false);
	});
});
