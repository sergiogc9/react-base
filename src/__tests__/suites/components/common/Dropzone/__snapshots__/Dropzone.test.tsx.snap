// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test dropzone component Test files render 1`] = `"<div id=\\"dropzoneUploadedFiles\\"><div class=\\"dropzone-upload-files-inner\\"><div class=\\"dropzone-uploaded-file\\"><div class=\\"dropzone-uploaded-file-info\\"><div class=\\"dropzone-uploaded-file-name\\">test_file.png</div></div><div class=\\"dropzone-uploaded-file-commands\\"><button id=\\"dropzoneUploadedFile0RemoveButton\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block dropzone-uploaded-file-commands-remove\\"><div class=\\"md-ink-container\\"></div><div class=\\"md-icon-separator\\"><i class=\\"md-icon material-icons md-text--inherit\\">close</i><span class=\\"md-icon-text\\"></span></div></button></div></div><div class=\\"dropzone-uploaded-file\\"><div class=\\"dropzone-uploaded-file-info\\"><div class=\\"dropzone-uploaded-file-name\\">test_extra_file.mp4</div></div><div class=\\"dropzone-uploaded-file-commands\\"><button id=\\"dropzoneUploadedFile1RemoveButton\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block dropzone-uploaded-file-commands-remove\\"><div class=\\"md-ink-container\\"></div><div class=\\"md-icon-separator\\"><i class=\\"md-icon material-icons md-text--inherit\\">close</i><span class=\\"md-icon-text\\"></span></div></button></div></div></div></div>"`;
